programs/server/Server.cpp

```diff
@@ -260,9 +263,62 @@ void checkForUsersNotInMainConfig(
 }
 
 
+void forkAndWatch(Poco::Logger * log)
+{
+    std::string original_process_name = process_name;
+
+    memset(process_name, 0, original_process_name.size());
+    strncpy(process_name, "clickhouse-watchdog", original_process_name.size());
+
+    setThreadName("clckhouse-watch");   /// 15 characters
+
+    while (true)
+    {
+        pid_t pid = fork();
+
+        if (-1 == pid)
+        {
+            LOG_FATAL(log, "Cannot fork");
+            exit(1);
+        }
+
+        if (0 == pid)
+        {
+            strncpy(process_name, original_process_name.data(), original_process_name.size());
+            setThreadName("clickhouse-serv");
+            return;
+        }
+
+        int status = 0;
+        if (-1 == waitpid(pid, &status, 0))
+        {
+            LOG_FATAL(log, "Cannot waitpid");
+            exit(2);
+        }
+
+        if (WIFEXITED(status))
+        {
+            LOG_INFO(log, "Child process exited normally with code {}.", WEXITSTATUS(status));
+            exit(status);
+        }
+
+        if (WIFSIGNALED(status))
+            LOG_FATAL(log, "Child process was terminated by signal {}.\n", WTERMSIG(status));
+        else if (WIFSTOPPED(status))
+            LOG_FATAL(log, "Child process was stopped by signal {}.\n", WSTOPSIG(status));
+        else
+            LOG_FATAL(log, "Child process was not exited normally by unknown reason.");
+
+        LOG_INFO(log, "Will restart.");
```

Comment:
So is clickhouse-server becoming pulseaudio now? A watchdog as a child process would be useful, but the reverse doesn't look good. For example, how am I supposed to learn the server pid to send it a signal? `$!` doesn't work anymore. And additional restart is just a really bad idea, the server must not reimplement process supervision in some non-standard way.

Reply:
@akuzm I don't want to leave automatic restart, there are some difficulties (for example, the `SYSTEM KILL` query will not have the desired effect of stopping the server).

Just to print message in log (stderr) and exit.

Reply:
> the server must not reimplement process supervision in some non-standard way

BTW I don't care about some standard ways at all.

Reply:
Watchdog itself should also be able to react signals (forward them to child process when needed)

Reply:
> Watchdog itself should also be able to react signals (forward them to child process when needed)

Watchdog should be a child process that prints a line to the log. This is simpler and has better usability than additional forking, while achieving the same result.

Reply:
@akuzm Parent process knows termination reason of child process but not vice versa.
For example, if you run a program in bash and it's killed by KILL signal, bash will print `Killed`.
We can achieve similar behaviour even if clickhouse-server is run as daemon.

Reply:
You're right, I confused it with Windows where you can wait for an arbitrary process to terminate. Linux added some new API in 5.3 for this, but it's mostly to avoid races with reusing pid, it still only possible to wait for a child process. Apparently they do plan to support it later https://lwn.net/Articles/794707/

So to recap, what we lose with forking is:
* simplicity of scripting -- there is no simple way to know which pid is `clickhouse-server` you just launched. Now I just use `$!`. For signals, we can implement forwarding, but this doesn't work if you need to inspect `/proc/pid`. For example, in perf tests I do this to collect memory maps of the server processes. 

* simplicity of managing the service on various distros -- don't know much about that, but can imagine that having a single process is easier than having two of them.

* simplicity of implementation -- we have to implement forking and signal forwarding. It is easy but still can be a source of bugs.

* `gdb`  doesn't follow child processes by default -- have to specially configure it. Other tools probably also break, e.g. `strace`, `perf`. 

What we get in exchange is a line about `clickhouse-server` killed by OOM killer, in the server log, so that we don't have to inspect `dmesg`.

Doesn't look like a great bargain to me, I'd much prefer to have the common tools work by default.

Reply:
I will enable it only when ClickHouse is run a daemon - already forked.
All cases with: run from terminal; run within gdb; run under service supervisor - will work as before.

> Doesn't look like a great bargain to me, I'd much prefer to have the common tools work by default.

I mostly agree... but 90% people don't know where to look when clickhouse-server just killed by signal 9.

Reply:
Actually if it's only for printing nice line in the log - it's not worth complicating imho. You can just write a line like 'previous server shutdown was not clear, please check dmsg' on process restart.

Reply:
We already have it.

Reply:
Why do you think it's not enough? 

Also about weird ideas - if we would have "some" process with oom_score_adj = 1000 it should be killed before clickhouse giving a chance to report smth (sounds hacky)

Reply:
Yes, I also had some thoughts about it, but it's unclear how much memory to allocate for "OOM canary".